src/hotspot/share/opto/loopopts.cpp

```diff
@@ -1503,19 +1503,32 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
           assert(!n_loop->is_member(get_loop(x_ctrl)), "should have moved out of loop");
           register_new_node(x, x_ctrl);
 
-          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr()) {
+          // Chain of AddP: (AddP base (AddP base )) must keep the same base after sinking. We don't add a CastPP here
+          // when the first one is sunk so if the second one is not, their bases remain the same. A CastPP of the base
+          // is only added once both AddP nodes are sunk (see special code for AddP after the cast is created).
+          if (x->in(0) == NULL && !x->is_DecodeNarrowPtr() &&
+              !(x->is_AddP() && x->in(AddPNode::Address)->is_AddP() && x->in(AddPNode::Address)->in(AddPNode::Base) == x->in(AddPNode::Base))) {
```

Comment:
I read the last check as "not a case of chained AddP with different bases". 
Then the comparison should be `!=`, shouldn't it?

Reply:
It's safe to add a cast if it's not a chained AddP with the same base otherwise there's a risk that adding a cast would make the base differ.

Reply:
Got it. Wasn't attentive enough when reading the description and the comment.